LFS-282: Searching of nested answers (not just direct children); LFS-254: As a user, I can see how my query matched each resource in the quick search results

IntegralProgrammer opened the pull request:
This PR re-implements the `quickSearch()` method so that *XPATH* is used instead of *JCR-SQL2*. The *XPATH* query is written so that not only direct children of the form but also nested answers are searched. This PR also takes front-end code from the *LFS-254* PR, specifically for highlighting the matching text substring for a matching answer.

This PR implements everything required by *LFS-254* except for ensuring that it is the *first* matching answer that is displayed in the auto-suggest results - this has been moved to a new sprint. Due to the large overlap in functionality between this PR and *LFS-254*, merging this PR will not allow for *LFS-254* to be merged over-top.

IntegralProgrammer committed:
Add queryXPATH() method

Modify quickSearch() to use queryXPATH() instead of queryJCR()

IntegralProgrammer committed:
Properly finds answers that are form children, sub-children, and so forth but the answer returned is not necessarily the first match on the questionnaire

IntegralProgrammer committed:
Remove debug logging

IntegralProgrammer committed:
Pull in JSX front-end code from LFS-254 so that a matching form section is highlighted

IntegralProgrammer committed:
Move shortening of long context matches from the front-end to the back-end

marta- commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/QueryBuilder.java:
```diff
@@ -156,6 +161,23 @@ private String fullTextEscape(String input)
+    {
+        String questionNodeId = res.getValueMap().get("question", "");
+        if (!"".equals(questionNodeId))
+        {
```
Our code style is to put the open curly bracket on a new line only for functions. Make this style change throughout your code.

```suggestion
        if (!"".equals(questionNodeId)) {
```

IntegralProgrammer replied:
Changed in commit `79ffe93` :heavy_check_mark:

marta- commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/QueryBuilder.java:
```diff
@@ -50,6 +53,8 @@
  */
 public class QueryBuilder implements Use
 {
+    private final int maxContextMatch = 8;
```
Code style:

```suggestion
    private final int MAX_CONTEXT_MATCH = 8;
```

IntegralProgrammer replied:
Changed in commit `88cae6c` :heavy_check_mark:

marta- commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/QueryBuilder.java:
```diff
@@ -167,25 +189,65 @@ private String fullTextEscape(String input)
+        {
+            Resource thisResource = foundResources.next();
+            Resource thisParent = thisResource;
+            String resourcevalue = thisResource.getValueMap().get("value", "");
```
Code style:
```suggestion
            String resourceValue = thisResource.getValueMap().get("value", "");
```

IntegralProgrammer replied:
Fixed in commit `818abd8` :heavy_check_mark:

IntegralProgrammer committed:
cast value property of answer node to an array of strings to properly deal with multi-valued properties

marta- commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/QueryBuilder.java:
```diff
@@ -165,27 +191,84 @@ private String fullTextEscape(String input)
+                    break;
+                }
+                thisParent = thisParent.getParent();
+            }
```
```suggestion
            while (thisParent != null && !"lfs/Form".equals(thisParent.getResourceType())) {
                thisParent = thisParent.getParent();
            }
```

IntegralProgrammer committed:
Refactor to use a newly created getMatchingFromArray() method

Remove debugging statements

IntegralProgrammer committed:
Rename maxContextMatch to MAX_CONTEXT_MATCH

IntegralProgrammer committed:
Update modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/QueryBuilder.java

Simplify code which moves upwards through the tree searching for the first found ancestor node of type "lfs/Form"

Co-Authored-By: Marta Girdea <marta.girdea@gmail.com>

IntegralProgrammer committed:
Fix naming convention of variables

IntegralProgrammer committed:
Fix brackets to conform to coding style rules

marta- commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/QueryBuilder.java:
```diff
@@ -167,25 +208,55 @@ private String fullTextEscape(String input)
      */
     private Iterator<Resource> quickSearch(String query) throws RepositoryException
     {
-        final String[] toSearch = {"lfs:Form", "lfs:Subject", "lfs:Questionnaire"};
```
FTR, this also seems to provide the quick fix for [LFS-323: Quick search matches found in questionnaires are irrelevant](https://phenotips.atlassian.net/browse/LFS-323).

Is everyone on board with removing subjects and questionnaires from quick search? @OrdiNeu @sdumitriu

marta- replied:
Addressed in commit 87559a8

sashaandjic commented:
Problem when searching '%', '_' and '"'.
![image](https://user-images.githubusercontent.com/23243244/75584190-33fdd700-5a3d-11ea-914e-a6169fce7ea6.png)

OrdiNeu committed:
LFS-282: Searching of nested answers (not just direct children)
Bugfix on the query erroring out on wildcard characters (%, _) and also
double quotation marks (")

OrdiNeu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/QueryBuilder.java:
```diff
@@ -167,25 +208,55 @@ private String fullTextEscape(String input)
+                thisParent = thisParent.getParent();
+            }
+            if (thisParent != null) {
+                String resourceValue = getMatchingFromArray(resourceValues, query);
```
```suggestion
            String[] resourceValues = thisResource.getValueMap().get("value", String[].class);
            String resourceValue = getMatchingFromArray(resourceValues, query);
            while (thisParent != null && !"lfs/Form".equals(thisParent.getResourceType())) {
                thisParent = thisParent.getParent();
            }
            if (thisParent != null && resourceValue != null) {
```
Guard against `getMatchingFromArray` returning `null` by moving the resourceValue outside of the conditional and ensuring it is not null (e.g. when the search was found using wildcard characters)

IntegralProgrammer committed:
Update modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/QueryBuilder.java

Guard against getMatchingFromArray returning null

Co-Authored-By: OrdiNeu <OrdinaryNeumann@gmail.com>

OrdiNeu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/QueryBuilder.java:
```diff
@@ -153,7 +159,42 @@ private String fullTextEscape(String input)
+        if (!"".equals(questionNodeId)) {
+            Node questionNode = res.adaptTo(Node.class).getSession().getNodeByIdentifier(questionNodeId);
+            return questionNode.getProperty("text").getValue().toString();
+        }
```
It is much easier in this case to just access the node directly, rather than obtaining the session and obtaining the node by ID.
```suggestion
        Node questionNode = res.adaptTo(Node.class).getProperty("question").getNode();
        if (questionNode != null) {
            return questionNode.getProperty("text").getString();
        }
```

IntegralProgrammer replied:
I have addressed this issue, and other robustness fixes, in commit `61bac50`. :heavy_check_mark:

OrdiNeu reviewed (approved):
Few quick changes to address, otherwise looks good :heavy_check_mark:

IntegralProgrammer committed:
To prevent accidental collisions, the key queryMatch has been replaced with the key lfs:queryMatch and assigned to the constant LFS_QUERY_MATCH_KEY

IntegralProgrammer commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/QueryBuilder.java:
```diff
@@ -167,25 +208,55 @@ private String fullTextEscape(String input)
+                };
+                Node thisParentNode = thisParent.adaptTo(Node.class);
+                if (!thisParentNode.hasProperty("queryMatch")) {
+                    thisParentNode.setProperty("queryMatch", queryMatch);
```
I like this idea - I have implemented it in commit `d27a912`. :heavy_check_mark:

IntegralProgrammer committed:
Simplify getQuestion() method and make it fail gracefully if 'question' or 'text' properties are not present

OrdiNeu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/QueryBuilder.java:
```diff
@@ -172,12 +172,15 @@ private String fullTextEscape(String input)
      */
     private String getQuestion(Resource res) throws ItemNotFoundException, RepositoryException
```
```suggestion
    private String getQuestion(Resource res) throws ItemNotFoundException
```
If we're catching the `RepositoryException` in this function, it's no longer necessary to re-throw it.

IntegralProgrammer replied:
Done - see commit `5eb1a09` :heavy_check_mark:

IntegralProgrammer committed:
Remove throwing of RepositoryException which is always caught

marta- committed:
LFS-282: Some refactoring of the code for displaying the match results + added the ability to style the match key, before and after text

sashaandjic commented:
% is still causing error
![image](https://user-images.githubusercontent.com/23243244/75721259-17180c80-5ca6-11ea-8160-8af38f5e4f99.png)

sashaandjic commented:
Tested.
Works as expected.

marta- committed:
LFS-282: Removed UI elements dealing with resource types that are no longer supported by the quick search backend (Questionnaires and Subjects)

sashaandjic reviewed (approved):
tested

marta- committed:
LFS-282: Fixed regressions introduced by commit 87559a8
- "Searching..." and "No results" messages were no longer displayed,
- Resource id not fully visible as id no longer wraps

sashaandjic reviewed (approved):
tested again
